### PR TITLE
[Quant][fx] Define native backend_config_dict for linear and conv

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74636

Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35225680](https://our.internmc.facebook.com/intern/diff/D35225680)

### CODE REVIEW

File: torch/ao/quantization/fx/backend_config/native.py [hunk 1]
```diff
@@ -1,44 +1,216 @@
...
+            "pattern": convs.bn_relu_qat,
+            "observation_type": observation_type,
+            "dtype_configs": dtype_configs,
+            "root_module": convs.root,
+            "reference_quantized_module_for_root": convs.reference,
+        })
+    return conv_configs
+
+def get_native_backend_config_dict():
+    """ Get backend for PyTorch Native backend_config_dict (fbgemm/qnnpack). """
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/backend_config/native.py:208]: nit: swap `backend` and `backend_config_dict`?

**Review by jerryzh168 — APPROVED** [review]
looks great, thanks!

**Comment by jerryzh168** [comment]
there is some error in numeric suite, which I already fixed in my PR, should I land mine first? it might have some merge conflict with your PR

**Comment by andrewor14** [comment]
Yes that's fine. I was expecting conflicts when I moved the dtype_config

File: torch/ao/quantization/fx/match_utils.py [hunk 2]
```diff
@@ -124,16 +124,21 @@ def record_match(pattern, node, matched):
         else:
             matched.append(node)
 
+    # Ensure tuple patterns are matched first, e.g. match (F.relu, F.linear) before F.relu,
+    # such that we do not need to impose an ordering requirement on the patterns map
+    sorted_patterns = sorted(patterns.keys(), key=lambda k: -1 if isinstance(k, tuple) else 1)
```
Comment (andrewor14) [inline torch/ao/quantization/fx/match_utils.py:129]: This was needed to fix functional linear/conv + relu

Comment (andrewor14) [inline torch/ao/quantization/fx/match_utils.py:129]: This was needed to fix functional linear/conv + relu

File: torch/ao/quantization/fx/backend_config/native.py [hunk 3]
```diff
@@ -92,16 +96,205 @@ def _get_default_op_backend_config(op, dtype_configs):
...
-    """ Get backend for PyTorch Native backend_config_dict (fbgemm/qnnpack)
+def _get_linear_configs():
     """
+    Return all configs related to linear modules and ops.
+    """
+    observation_type = ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT
+    dtype_configs = [weighted_op_int8_dtype_config]
+    linear_configs = []
+
+    # linear module
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/backend_config/native.py:107]: can we classify (by adding comments) these configs a bit to make sure we have more structured support,
e.g.
1. single linear
2. linear + relu
3. linear + bn

File: torch/ao/quantization/fx/backend_config/native.py [hunk 4]
```diff
@@ -92,16 +96,205 @@ def _get_default_op_backend_config(op, dtype_configs):
...
+    linear_configs.append({
+        "pattern": nniqat.LinearBn1d,
+        "observation_type": observation_type,
+        "dtype_configs": dtype_configs,
+        "root_module": torch.nn.Linear,
+        "reference_quantized_module_for_root": nnqr.Linear,
+    })
+    return linear_configs
+
+def _get_conv_configs():
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/backend_config/native.py:196]: same for conv I think, we can classify these configs a bit, to make sure it is easier to add things in the future.

**Review by jerryzh168 — APPROVED** [review]
Looks great! had some nit comments inline

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

File: torch/ao/quantization/fx/backend_config/native.py [hunk 5]
```diff
@@ -92,16 +96,222 @@ def _get_default_op_backend_config(op, dtype_configs):
...
+    """ Get backend_config_dict for PyTorch Native backend (fbgemm/qnnpack). """
     return {
         # optional
         "name": "native",
         "configs": [
-            _LINEAR_MODULE_CONFIG,
             *_DEFAULT_OP_INT8_CONFIGS,
             *_DEFAULT_OP_FP16_CONFIGS,
             *_DEFAULT_OP_INT8_OR_FP16_CONFIGS,
+            *_get_linear_configs(),
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/backend_config/native.py:314]: I think we also support fp16 linear quantization, how did we support them here in the config?

Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:314]: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/_lower_to_native_backend.py#L286

Comment (andrewor14) [reply torch/ao/quantization/fx/backend_config/native.py:314]: (As discussed offline, these will be covered in a future PR)

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

**Pull request closed without merging**